psi4/src/psi4/libmints/writer.cc

```diff
@@ -318,19 +316,59 @@ void FCHKWriter::write(const std::string &filename) {
     chk_ = fopen(filename.c_str(), "w");
     std::shared_ptr<BasisSet> basis = wavefunction_->basisset();
     int maxam = basis->max_am();
-    if (maxam > 4) throw PSIEXCEPTION("The Psi4 FCHK writer only supports up to G functions");
     std::shared_ptr<Molecule> mol = wavefunction_->molecule();
+    std::shared_ptr<Wavefunction> refwf = wavefunction_->reference_wavefunction();
+
+    // Orbitals
     SharedMatrix Ca_ao = wavefunction_->Ca_subset("AO");
     SharedMatrix Cb_ao = wavefunction_->Cb_subset("AO");
-    SharedMatrix Da_ao = wavefunction_->Da_subset("AO");
-    SharedMatrix Db_ao = wavefunction_->Db_subset("AO");
-    SharedMatrix Dtot_ao(Da_ao->clone());
-    SharedMatrix Dspin_ao(Da_ao->clone());
-    const std::string &name = wavefunction_->name();
-    const std::string &basisname = basis->name();
+
+    // SCF density matrices
+    SharedMatrix Da_ao;
+    SharedMatrix Db_ao;
+    // Post-Hartree-Fock density matrices
+    SharedMatrix DPHFa_ao;
+    SharedMatrix DPHFb_ao;
+
+    // Post Hartree Fock?
+    bool pHF = (refwf != NULL);
+
+    if (pHF) {
+        // Level of theory may be correlated
+        Da_ao = refwf->Da_subset("AO");
+        Db_ao = refwf->Db_subset("AO");
+
+        DPHFa_ao = wavefunction_->Da_subset("AO");
+        DPHFb_ao = wavefunction_->Db_subset("AO");
+    } else {
+        // SCF level of theory
+        Da_ao = wavefunction_->Da_subset("AO");
+        Db_ao = wavefunction_->Db_subset("AO");
+    }
+
+    // Total and spin density
+    SharedMatrix Dtot_ao;
+    SharedMatrix Dspin_ao;
+    SharedMatrix DPHFtot_ao;
+    SharedMatrix DPHFspin_ao;
+
+    Dtot_ao = Da_ao->clone();
     Dtot_ao->add(Db_ao);
+    Dspin_ao = Da_ao->clone();
     Dspin_ao->subtract(Db_ao);
+
+    if (pHF) {
+        DPHFtot_ao = DPHFa_ao->clone();
+        DPHFtot_ao->add(DPHFb_ao);
+        DPHFspin_ao = DPHFa_ao->clone();
+        DPHFspin_ao->subtract(DPHFb_ao);
+    }
+
+    const std::string &name = wavefunction_->name();
+    const std::string &basisname = basis->name();
     int nbf = basis->nbf();
+    assert(Ca_ao->nrow() == nbf);
```

Comment:
Any reason for the assert here, by definition within Psi4 this is true.

Reply:
I guess I wanted to make sure in case Ca_ao was symmetry blocked. Removed